### PR TITLE
DATA: Contributor 

### LINKED ISSUE #5: DATA: Contributor
## Model: Contributor

Properties:
| Field       | Type                             | Description                                        |
|-------------|----------------------------------|----------------------------------------------------|
| Id          | Guid                           | Our internal identifier                            |
| ExternalId  | string (Required, max 255)     | The external unique identifier                     |
| SourceId      | Guid (Required)     | The source Id of the source     |
| Username    | string (Required, max 255)     | The username / login of the contributor            |
| Name        | string (Optional, max 255)     | The name of the contributor                        |
| Email       | string (Optional, max 255)     | The email address of the contributor               |
| AvatarUrl   | string (Optional)              | The avatar Url of the contributor                  |
| CreatedAt     | DateTimeOffset         | When the remote repo was created     |
| UpdatedAt     | DateTimeOffset         | When the remote repo was last updated    |

** **_Extra considerations: Add a unique constraint on (ExternalId, Source). This will ensure we do not have duplicate contributors for a source._**

image

### PR DESCRIPTION
**Pull request by Lezune**
Closes #5

### CODE REVIEW

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using GitFyle.Core.Api.Models.Foundations.Contributor;
```
Comment (glhays) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.cs:1]: Add copyright header please.

**Review by glhays — COMMENTED** [review]
New AddContributorProperties migration needed to add the two new properties.

**Comment by Lezune** [comment]
> New AddContributorProperties migration needed to add the two new properties.

Done.

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Missing the fluent configuration and the registration of that configuration on the StorageBroker

**Comment by Lezune** [comment]
> > This has been resolved.
> 
> I don' think this has been resolved. I can't see the the file `StorageBroker.Contributors.Configuration.cs` Can you double check that this file has been committed with the fluent configuration?
> 
> image
> 
> Have a look at https://github.dev/The-Standard-Organization/GitFyle.Core.Api/pull/32 if you need a sample.
> 
> You will also need to add this to the StorageBroker.cs:
> 
> ```cs
>         protected override void OnModelCreating(ModelBuilder modelBuilder)
>         {
>             AddContributorConfigurations(modelBuilder.Entity<Contributor>());
>         }
> ```

@cjdutoit Sorry, I misunderstood what was being requested. I believe the most recent edit has what you are looking for but please review to confirm issue has been resolved.

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+            });
+
+            builder.HasOne(x => x.Source)
+                .WithMany(x => x.Contributors)
+                .HasForeignKey(x => x.SourceId)
+                .OnDelete(Microsoft.EntityFrameworkCore.DeleteBehavior.NoAction);
+
+        }
+
```
Comment (glhays) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs:47]: Remove extra line

Comment (Lezune) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs]: This has been resolved.

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            builder.Property(contributor => contributor.Email)
+                .HasMaxLength(255);
+
+            builder.HasIndex(contributor => new
+            {
+                    contributor.ExternalId,
+                    contributor.SourceId,
+                    contributor.Username,
+
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs:38]: Remove extra line

Comment (Lezune) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs]: This has been resolved.

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
+﻿// ----------------------------------------------------------------------------------
+// Copyright (c) The Standard Organization: A coalition of the Good-Hearted Engineers
+// ----------------------------------------------------------------------------------
+
+using GitFyle.Core.Api.Models.Foundations.Contributors;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs:6]:
```suggestion
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                    contributor.SourceId,
+                    contributor.Username,
+            });
+
+            builder.HasOne(x => x.Source)
+                .WithMany(x => x.Contributors)
+                .HasForeignKey(x => x.SourceId)
+                .OnDelete(Microsoft.EntityFrameworkCore.DeleteBehavior.NoAction);
+        }
+    }
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs:45]: Always give your lambda parameters a meaningfull name.  It helps with readability and it acts as a form of documentations as it explains what each part of the expression is supposed to represent without needing additional comments.

```suggestion
            builder.HasOne(contributor => contributor.Source)
                .WithMany(source => source.Contributors)
                .HasForeignKey(contributor => contributor.SourceId)
                .OnDelete(DeleteBehavior.NoAction);
```

Comment (Lezune) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs]: Applied changes. Please review to confirm issue has been resolved.

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+            builder.HasIndex(contributor => new
+            {
+                    contributor.ExternalId,
+                    contributor.SourceId,
+                    contributor.Username,
+            });
+
+            builder.HasOne(contributor => contributor.Source)
+                .WithMany(contributor => contributor.Contributors)
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs:42]:
```suggestion
                .WithMany(source => source.Contributors)
```

File: GitFyle.Core.Api/Models/Foundations/Contributors/Contributor.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    public class Contributor 
+    { 
+        public Guid Id { get; set; }
+        public string ExternalId { get; set; }
+        public Guid SourceId { get; set; }
+        public string Username { get; set; }
+        public string Name { get; set; }
+        public string Email {  get; set; }
+        public string AvatarUrl { get; set; }
+        public Source Source { get; set; }
```
[the remark this answers is not in the document]
Comment (Lezune) [reply GitFyle.Core.Api/Models/Foundations/Contributors/Contributor.cs]: This has been updated. Please review to confirm this issue has been resolved.

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs (hunk 6 above)
Comment (Lezune) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributors.Configuration.cs]: This has been updated. Please review to confirm this issue has been resolved.

File: GitFyle.Core.Api/Models/Foundations/Contributors/Contributor.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lezune) [reply GitFyle.Core.Api/Models/Foundations/Contributors/Contributor.cs]: This has been updated. Please review to confirm this issue has been resolved.

**Review by cjdutoit — APPROVED** [review]

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Please resolve the conflicts by pulling the latest code from main.  You will also need to re-apply your migrations after the merge

**Comment by Lezune** [comment]
Hello @cjdutoit. I noticed something in the StorageBrokerModelSnapshot that seemed to be off. I wanted to double check with you before making any edits to the code. I've noticed that there are a few properties in the Repository section that are missing(Id, CreatedAt, UpdatedAt). Am I fine with adding these in? It is causing an error when I try to re-apply the migration.

**Comment by Lezune** [comment]
> Hello @cjdutoit. I noticed something in the StorageBrokerModelSnapshot that seemed to be off. I wanted to double check with you before making any edits to the code. I've noticed that there are a few properties in the Repository section that are missing(Id, CreatedAt, UpdatedAt). Am I fine with adding these in? It is causing an error when I try to re-apply the migration.

Looks like the Id property was removed from the StorageBrokerModelSnapshot in commit 5a20f53.

**Comment by hassanhabib** [comment]
@Lezune pull latest from main and run the migration on a clean DB and see if everything works fine. If not, then delete all migration files and regenerate.

**Comment by Lezune** [comment]
> @Lezune pull latest from main and run the migration on a clean DB and see if everything works fine. If not, then delete all migration files and regenerate.

@hassanhabib Thank you for getting back to me on that. Was hesitant to delete anything outside my code but i followed your suggestion and now it has been fixed.

**Comment by Lezune** [comment]
> Please resolve the conflicts by pulling the latest code from main. You will also need to re-apply your migrations after the merge

This has been resolved.

**Pull request merged**